python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,296 @@
+        the reconciler/instance launcher will update the instance status
+        based on the ray status and the instance status with the underlying
+        node provider.
+
+    For full status transitions, see:
+    https://docs.google.com/document/d/1NzQjA8Mh-oMc-QxXOa529oneWCoA8sDiVoNkBqqDb4U/edit#heading=h.k9a1sp4qpqj4
+    """
+
+    def __init__(
+        self,
+        instance_storage: InstanceStorage,
+        available_node_types: Dict[str, Any],
+        status_change_subscribers: Optional[List[InstanceUpdatedSubscriber]] = None,
+    ) -> None:
+        super().__init__()
+        self._instance_configs = available_node_types
+        self._instance_storage = instance_storage
+        instance_storage.add_status_change_subscribers(status_change_subscribers)
+
+    def update_instance_manager_state(
+        self, request: UpdateInstanceManagerStateRequest
+    ) -> UpdateInstanceManagerStateReply:
+        """
+        Updates the instance manager state with the updated instances, and new
+        instances to launch.
+
+        Overrides:
+            Overrides the base class method.
+        """
+
+        # 1. Handle instance status updates.
+        ids_to_updates = {update.instance_id: update for update in request.updates}
+        if len(ids_to_updates) == 0:
+            to_update_instances, version = {}, request.expected_version
+        else:
+            to_update_instances, version = self._instance_storage.get_instances(
+                ids_to_updates.keys()
+            )
+            if request.expected_version >= 0 and request.expected_version != version:
+                reply = UpdateInstanceManagerStateReply()
+                reply.status = Status.OK
+                reply.version = version
+                return reply
+
+        for instance in to_update_instances.values():
+            update = ids_to_updates[instance.instance_id]
+            InstanceUtil.set_status(instance, new_ray_status=update.new_ray_status)
+
+        # 2. Handle new instances to start
+        new_instances = []
+        for launch_request in request.launch_requests:
+            instances = self._create_instances(launch_request)
+            new_instances.extend(instances)
+
+        expected_version = (
+            request.expected_version if request.expected_version else None
+        )
+
+        # 3. Update the instance storage.
+        ok, version = self._instance_storage.batch_upsert_instances(
+            new_instances + list(to_update_instances.values()),
+            expected_version,
+        )
+
+        if ok:
+            reply.status = Status.OK
+        else:
+            if version != expected_version:
+                reply.status = Status.VERSION_MISMATCH
+            else:
+                reply.status = Status.UNKNOWN_ERROR
+
+        reply = UpdateInstanceManagerStateReply()
+        reply.state.CopyFrom(self.get_instance_manager_state().state)
```

Comment:
// the updated instance state if success, else none.

Seems like it doesn't follow this specification when status != ok

Reply:
good catch.